### PR TITLE
92/Refactor pressure level modifications

### LINKED ISSUE #92: Refactor: `cubewrite()` pressure level correction to tested function
Tasks:
* [x] Extract this block https://github.com/ACCESS-NRI/um2nc-standalone/blob/main/umpost/um2netcdf.py#L150-L160
* [x] Add unit tests (~2)
* [ ] Move into `process()`

### PR DESCRIPTION
**Pull request by truth-quark**
Resolves #92.

This PR covers a few tasks:
* extracts pressure level modification functionality to a standalone function
* adds partial unit testing
* rearranges the `um2netcdf` module slightly
* partial code cleanup

**Known gaps:**

This work exposes a (possibly one off) case where `iris` funcs return a _new_ cube. Moving the pressure level fixes into `process()` requires modifying the list of cubes being processed to patch the new cube in. This is noted in the code. The tests are also moderately loose, related to the complexity with testing `iris` components.

It's likely addressing this will have to be deferred until the code is restructured & tested, to allow workflow refactoring.

Otherwise, any comments welcome!

### CODE REVIEW

**Review by CodeGat — APPROVED** [review]
Looks logical to me.

File: test/test_um2netcdf.py
Comment (truth-quark) [inline test/test_um2netcdf.py:693]: The fixture below should be replaceable with the `DummyCubeWithCoords` from the lat/long fixes.

File: umpost/um2netcdf.py
Comment (marc-white) [inline umpost/um2netcdf.py:760]: Is it possible to ever get a cube where the pressure data isn't monotonic for some reason (including user stupidity)?

File: test/test_um2netcdf.py
Comment (marc-white) [inline test/test_um2netcdf.py:764]: Should the `None` return be asserted here?

File: test/test_um2netcdf.py
Comment (blimlim) [inline test/test_um2netcdf.py:802]: Is it worth checking whether the mocked `iris.util.reverse()` is called here, or does that not really add anything?

**Review by blimlim — APPROVED** [review]
Changes look good to me. One small question but not really a necessary change

File: test/test_um2netcdf.py
Comment (blimlim) [reply test/test_um2netcdf.py:802]: Sounds like a good idea!

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:760]: I'm not sure if the data is always monotonic. 

In the "aiihca.paa1jan" dataset (`gadi:/g/data/tm70/esm15_testing/sample_output/aiihca.paa1jan`), extracting all the pressure coords from 100+ cubes gives 11 copies of these points for `pressure`:

```python
array([   1.,    5.,   10.,   20.,   30.,   50.,   70.,  100.,  150.,
         200.,  250.,  300.,  400.,  500.,  600.,  700.,  850.,  925.,
        1000.])
```
@MartinDix do you know if there is any guarantee that the pressure data is monotonic?

Comment (truth-quark) [reply umpost/um2netcdf.py:760]: Let's throw out an idea for stepwise simplification. The 1st goal for `um2nc-standalone` is to modularise & retrofit unit tests to the code. Changing/extending behaviour is a secondary step, made easier once tests exist. Therefore, an option for this PR/feature is:

* (Optional) add a `TODO: add monotonic test func` reminder to the code
* Create GitHub issue specifically for addressing the monotonic test (do here, add data preprocessing script?)
* Verify this extracted func matches the existing behaviour
* Merge this smaller work unit

How does that sound?

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:764]: Yes it should, since the fix pressure levels calling code is relying on a `cube` or a value to resolve as `False`.

Good catch!

Comment (truth-quark) [reply test/test_um2netcdf.py]: Fixed in dd9c18dd.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:760]: Ooops, a `monotonic` check function is included in `iris`:
https://github.com/SciTools/iris/blob/main/lib/iris/util.py#L615

Comment (marc-white) [reply umpost/um2netcdf.py:760]: Do we know if that check function gets applied to enforce the data to be monotonic, or is it just something we can use in tests?

File: test/test_um2netcdf.py
Comment (truth-quark) [inline test/test_um2netcdf.py:805]: NB: this is a BUG, the pressure points should be reversed

File: umpost/um2netcdf.py
Comment (truth-quark) [inline umpost/um2netcdf.py:760]: I noticed line 760 is using the entire `cube` in reversing the pressure levels:
https://github.com/ACCESS-NRI/um2nc-standalone/blob/dd9c18dd1ad709bd7c3dd839d3eb74bc6ebbfae4/umpost/um2netcdf.py#L757-L760

However, `iris.util.reverse()` takes `cube`s or `array`s:
https://github.com/SciTools/iris/blob/e1f5a8b880e443c58cba7770a1ffd32e067a64f2/lib/iris/util.py#L518-L524 (preview is failing....)

I'm experimenting with modifying the pressure array order in place, in the event the coord can be updated within the `cube`. It's currently breaking...

@MartinDix from your knowledge of `iris` do you know if modifying the `pressure.points` array is recommended practice?

```python
    pressure = cube.coord('pressure')

    # snip processing...

    if pressure.points[0] < pressure.points[-1]:
        # Flip to get pressure decreasing as per CMIP6 standard
        # return iris.util.reverse(cube, 'pressure')  # ORIG
        pressure.points = np.flip(pressure.points)  # TODO: is this valid???
```

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:760]: `monotonic()` isn't called directly in `iris.util.reverse()`:
https://github.com/SciTools/iris/blob/e1f5a8b880e443c58cba7770a1ffd32e067a64f2/lib/iris/util.py#L518-L612

File: umpost/um2netcdf.py
Comment (blimlim) [reply umpost/um2netcdf.py:760]: Just to jump in with some extra info, let me know if I've misunderstood the question though! My understanding is that the pressure coordinate's points ordering shouldn't change separately from of the rest of the cube's data.

My (rough) understanding of a typical `iris.cube` is that it holds an n-dimensional array as it's `data` attribute, and for each dimension has a separate coordinate array. The points in the coordinate arrays specify the locations corresponding to the main `data`. E.g. a cube describing surface temperature might have a 180x360 array of values, plus a length 180 `latitude` coordinate array, and length 360 longitude coordinate array. 

The temperature given by `cube.data[10,55]` would correspond to the location specified by `cube.coord("latitude")[10]` and `cube.coord("latitude")[55]`. But this means if we flip one of the coordinates independently of the data, the wrong locations will be given for the cube data.

It looks like `iris.util.reverse(cube, 'pressure')` is flipping both the `pressure` coordinate, and the actual cube's `data`, so that the two still line up:

```python
> cube_on_plev = cubes[142] # Get a cube with a pressure coordinate

# extract single latitude and longitude point, so that vertical coordinate pressure is the only coordinate
> single_cell_cube = cube_on_plev.extract(iris.Constraint(latitude=0.625, longitude = 0.9375))

# Print out initial pressure points and data array
> single_cell_cube.coord("pressure").points

array([   1.,    5.,   10.,   20.,   30.,   50.,   70.,  100.,  150.,
        200.,  250.,  300.,  400.,  500.,  600.,  700.,  850.,  925.,
       1000.])

> single_cell_cube.data
array([2.23973939e-04, 1.04892252e-02, 5.33754104e-02, 2.66307022e-01,
       9.27435023e-01, 6.87145581e+00, 3.06664567e+01, 5.38439087e+01,
       5.30945653e+01, 4.83872854e+01, 2.50941701e+01, 1.22997686e+01,
       9.74147291e+00, 2.95951569e+01, 5.42121837e+01, 5.50435870e+01,
       5.61263122e+01, 8.55014709e+01, 7.11272948e+01])

# Reverse cube according to pressure coordinate
> single_cell_cube_reverse = iris.util.reverse(single_cell_cube, single_cell_cube.coord("pressure"))

# Print out new data and pressure coordinate arrays
> single_cell_cube_reverse.coord("pressure").points
array([1000.,  925.,  850.,  700.,  600.,  500.,  400.,  300.,  250.,
        200.,  150.,  100.,   70.,   50.,   30.,   20.,   10.,    5.,
          1.])

>single_cell_cube_reverse.data
array([7.11272948e+01, 8.55014709e+01, 5.61263122e+01, 5.50435870e+01,
       5.42121837e+01, 2.95951569e+01, 9.74147291e+00, 1.22997686e+01,
       2.50941701e+01, 4.83872854e+01, 5.30945653e+01, 5.38439087e+01,
       3.06664567e+01, 6.87145581e+00, 9.27435023e-01, 2.66307022e-01,
       5.33754104e-02, 1.04892252e-02, 2.23973939e-04])

# Both have flipped!
```

File: umpost/um2netcdf.py
Comment (marc-white) [reply umpost/um2netcdf.py:760]: The test condition here probably needs to be updated to use `monotonic()` then - until we can confirm the pressure values are guaranteed to be monotonic, the current test doesn't actually confirm the direction of the pressure values.

Comment (truth-quark) [reply umpost/um2netcdf.py:760]: > Do we know if that check function gets applied to enforce the data to be monotonic, or is it just something we can use in tests?

As a standalone function, we can use it in both cases.

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py]: Changed in 1941c693.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py]: Added TODO to consider monotonic checking in f6ba005c.

Comment (truth-quark) [reply umpost/um2netcdf.py]: I've deferred this to #107 to help bring the rest of the work into the `main` branch.

File: test/test_um2netcdf.py
Comment (truth-quark) [reply test/test_um2netcdf.py:695]: Issue #63 covers this task.

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:761]: Ok, that makes sense.

I've skipped the test on account of problems at the um2nc & iris "boundary". Implementing a mocked `reverse()` requires faffing with return data structures OR using the real function & making a complicated custom cube arg. Both options are ugly for testing a _one liner_ call to an external library.

I can leave the skipped test in for future work OR delete it & comment why the pressure reversal isn't tested. Does anyone (@blimlim @marc-white  have a preference either way?

**Comment by truth-quark** [comment]
This one needs a final review before merging. @blimlim @marc-white would either of you mind going over the latest changes & conversation threads?

File: umpost/um2netcdf.py
Comment (marc-white) [reply umpost/um2netcdf.py:761]: I think the skipped test is fine for now, but add an issue to deal with it later. It definitely needs to be tested, but data transformations (even ones as basic as flipping) are a massive faff that need to be stepped through very methodically.

**Review by marc-white — APPROVED** [review]
I think we're good to go at this point!

File: umpost/um2netcdf.py
Comment (truth-quark) [reply umpost/um2netcdf.py:761]: #108 now covers the testing task.

I think deferring is a good step for now, to free effort for modularising & retrofitting tests to other functions.

**Comment by truth-quark** [comment]
> I think we're good to go at this point!

Argh, GitHub seems to need someone else to resolve the conversations before it'll let me merge...

**Review by CodeGat — APPROVED** [review]

**Pull request merged**
